protocol/client/messenger_v2.go

```diff
@@ -32,33 +32,47 @@ type MessengerV2 struct {
 	db       Database
 
 	mu      sync.Mutex
-	public  map[string]AsyncStream
-	private map[string]AsyncStream
+	public  map[string]*Stream
+	private map[string]*Stream
 
 	events *event.Feed
 }
 
 func (m *MessengerV2) Start() error {
+	log.Printf("[Messenger::Start]")
+
 	m.mu.Lock()
 	defer m.mu.Unlock()
+
 	contacts, err := m.db.Contacts()
 	if err != nil {
 		return errors.Wrap(err, "unable to read contacts from database")
 	}
 
+	// For each contact, a new Stream is created that subscribes to the protocol
+	// and forwards incoming messages to the Stream instance.
+	// We iterate over all contacts, however, there are two cases:
+	// (1) Public chats where each has a unique topic,
+	// (2) Private chats where a single or shared topic is used.
+	// This means that we don't know from which contact the message
+	// came from until it is examined.
 	for i := range contacts {
 		options, err := createSubscribeOptions(contacts[i])
 		if err != nil {
 			return err
 		}
+
+		// For contacts with public key, we just need to make sure
+		// each possible topic has a stream. For a single topic
+		// for all private conversations, the map will have a len of 1.
 		if contacts[i].Type == ContactPublicKey {
 			_, exist := m.private[contacts[i].Topic]
 			if exist {
 				continue
 			}
-			stream := NewStream(context.Background(), options, m.proto, NewPrivateHandler(m.db))
-			err := stream.Start()
-			if err != nil {
+
+			stream := NewStream(m.proto, StreamHandlerMultiplexed(m.db))
```

Comment:
multiplexed is a bit too generic. i don't like private as well so idk

Reply:
`StreamHandlerForMultiplexedTopic`? Super long but it should describe exactly what it is -- handle messages for many contacts over a single topic.